Source/Core/InputCommon/ControllerInterface/Android/Android.cpp

```diff
@@ -1,32 +1,1163 @@
+      jobject motion_range = env->CallObjectMethod(motion_ranges_list, s_list_get, i);
+
+      jint source = env->CallIntMethod(motion_range, s_motion_range_get_source);
+      jint axis = env->CallIntMethod(motion_range, s_motion_range_get_axis);
+      jfloat min = env->CallFloatMethod(motion_range, s_motion_range_get_min);
+      jfloat max = env->CallFloatMethod(motion_range, s_motion_range_get_max);
+
+      env->DeleteLocalRef(motion_range);
+
+      AndroidAxis* positive = nullptr;
+      AndroidAxis* negative = nullptr;
+      if (max > 0)
+        positive = new AndroidAxis(source, axis, false);
+      if (min < 0)
+        negative = new AndroidAxis(source, axis, true);
+
+      if (positive && negative)
+        AddAnalogInputs(positive, negative);
+      else if (positive || negative)
+        AddInput(positive ? positive : negative);
+    }
+    env->DeleteLocalRef(motion_ranges_list);
+  }
+
+  jobject AddSensors(JNIEnv* env, jobject input_device)
+  {
+    jobject local_sensor_event_listener;
+    if (input_device)
+    {
+      local_sensor_event_listener =
+          env->NewObject(s_sensor_event_listener_class,
+                         s_sensor_event_listener_constructor_input_device, input_device);
+    }
+    else
+    {
+      local_sensor_event_listener =
+          env->NewObject(s_sensor_event_listener_class, s_sensor_event_listener_constructor);
+    }
+
+    jobject sensor_event_listener = env->NewGlobalRef(local_sensor_event_listener);
+
+    env->DeleteLocalRef(local_sensor_event_listener);
+
+    jobjectArray j_axis_names = reinterpret_cast<jobjectArray>(
+        env->CallObjectMethod(sensor_event_listener, s_sensor_event_listener_get_axis_names));
+
+    jbooleanArray j_negative_axes = reinterpret_cast<jbooleanArray>(
+        env->CallObjectMethod(sensor_event_listener, s_sensor_event_listener_get_negative_axes));
+    jboolean* negative_axes = env->GetBooleanArrayElements(j_negative_axes, nullptr);
+
+    const jsize axis_count = env->GetArrayLength(j_axis_names);
+    ASSERT(axis_count == env->GetArrayLength(j_negative_axes));
+    for (jsize i = 0; i < axis_count;)
+    {
+      AndroidSensorAxis* positive = nullptr;
+      AndroidSensorAxis* negative = nullptr;
+
+      if (!negative_axes[i])
+      {
+        const jstring axis_name =
+            reinterpret_cast<jstring>(env->GetObjectArrayElement(j_axis_names, i));
+        positive = new AndroidSensorAxis(env, sensor_event_listener, axis_name, false);
+        env->DeleteLocalRef(axis_name);
+        ++i;
+      }
+      // Intentional lack of `else`
+      if (i < axis_count && negative_axes[i])
+      {
+        const jstring axis_name =
+            reinterpret_cast<jstring>(env->GetObjectArrayElement(j_axis_names, i));
+        negative = new AndroidSensorAxis(env, sensor_event_listener, axis_name, true);
+        env->DeleteLocalRef(axis_name);
+        ++i;
+      }
+
+      if (positive && negative)
+        AddAnalogInputs(positive, negative);
```

Comment:
I'm basing this comment on `AddSensors` being strictly for accel/gyro. I'm not sure if there are other "sensors"?

The purpose of `AddAnalogInputs`, which should probably be more clearly named, is to add the provided inputs and additionally `FullAnalogSurface` versions of the inputs, which deal with e.g. analog triggers which are sometimes, depending on the gamepad, in their neutral state at a normalized `-1` or a `0`. The android input code, which I'm guessing is removed now, had some hard-coded special cases for dealing with this on xbox controllers and whatnot. It's unknowable to us how a trigger might be exposed so inputs for each situation are created. 

Anyways, this shouldn't be necessary for accel/gyro. We just need to `AddInput` both directions. Also, the whole `if (positive && negative)` check seems strange to me, particularly for the accel/gyro, but maybe I'm not digging deep enough. Shouldn't we always have accel neg and pos?

Reply:
> The purpose of `AddAnalogInputs`, which should probably be more clearly named, is to add the provided inputs and additionally `FullAnalogSurface` versions of the inputs, which deal with e.g. analog triggers which are sometimes, depending on the gamepad, in their neutral state at a normalized `-1` or a `0`.

I wasn't entirely sure what the actual purpose was, but that makes sense. Thanks for the information.

> The android input code, which I'm guessing is removed now, had some hard-coded special cases for dealing with this on xbox controllers and whatnot.

Yes, I removed that code.

> Anyways, this shouldn't be necessary for accel/gyro. We just need to `AddInput` both directions.

Yeah, with what you said in mind, I agree. I'll remove the use of `AddAnalogInputs` for sensors.

> Also, the whole `if (positive && negative)` check seems strange to me, particularly for the accel/gyro, but maybe I'm not digging deep enough. Shouldn't we always have accel neg and pos?

There are other sensors that have just a positive axis. For instance, the proximity sensor (originally intended for turning off the display if you hold your phone to your head during a phone call), the ambient light sensor, and the hinge angle (on a foldable phone). There's a full list in the `addSensors` method in DolphinSensorEventListener.java if you're interested. But indeed, for accel/gyro specifically, there's always both neg and pos.